Expose background subtraction options in GUI

gemmaguest opened the pull request:
Expose background subtraction options in GUI

Expose the background subtraction options in ReflectometryISISLoadAndProcess in the ISIS Reflectometry interface. The background processing instructions need to be specified on a per-angle basis so these are added to the Per-Angle Defaults table. The other options can be specified as defaults for all runs so are added as their own widgets on the Experiment Settings tab.

Fixes #27826

**To test:**

Requres the ISIS archive to be on.

- Open the `ISIS Reflectometry` interface
- Switch the instrument to `OFFSPEC` and enter run `53373` and angle `0.35` in the table.
- Click Process.
- Check the history of `IvsQ_binned_53373` and check that `SubtractBackground` is false and the other background inputs are default:
  - BackgroundCalculationMethod = PerDetectorAverage
  - DegreeOfPolynomial = 0
  - CostFunction = Least squares
- On the Experiment tab, tick SubtractBackground and then go back to the Runs tab and re-process. Check the history again and now SubtractBackground should be true with the others still defaults.
- Now change some of the other options and check their changes are reflected in the history.
- Also try entering processing instructions in the Per-angle defaults table, e.g. `380-389,416-426` and check that these appear in the history.
  - Try setting an angle in the Per-angle defaults table equal to `0.35` - it should be picked up in the history.
  - Try adding a new row with a different angle, e.g. `0.5` and set the Background. It should not be picked up in the history.
- Check that the background options can all be overriden by specifying properties manually in the Options field on the Runs table, e.g. `BackgroundProcessingInstructions='380-389,416-426'`
- From the Batch menu, save the settings to file and re-load them. Check they are all included when re-loaded.

**Release notes to follow**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

gemmaguest commented:
@giovannidisiena Have you made a start on this yet? If not, I was going to ask @rprospero to have a look because he's done some work on this.

giovannidisiena commented:
@gemmaguest only a quick first glance so feel free.

rprospero commented on qt/scientific_interfaces/test/ISISReflectometry/Reduction/ValidatePerThetaDefaultsTest.h:
```diff
@@ -132,6 +132,25 @@ class ValidatePerThetaDefaultsTest : public CxxTest::TestSuite {
+  void testParseBackgroundProcessingInstructionsError() {
+    PerThetaDefaultsValidator validator;
+    auto result = validator({"", "", "", "", "", "", "", "", "", "bad"});
+    std::vector<int> errorCells = {9};
```
Again, could `9` be replaced with `PerThetaDefaults::BACKGROUND_SPECTRA`?  Obviously this would also require including `PerThetaDefaults.h`

rprospero commented on qt/scientific_interfaces/ISISReflectometry/Reduction/BackgroundSubtraction.h:
```diff
@@ -0,0 +1,99 @@
+
+inline BackgroundSubtractionType
+backgroundSubtractionTypeFromString(std::string const &subtractionType) {
+  if (subtractionType.empty() || subtractionType == "PerDetectorAverage")
```
This might be a stylistic thing, but I was wondering if it might be worth creating a boost:bimap between the enum values and the strings?  Then most of these functions could be replaced with a simple value lookup.

gemmaguest replied:
Good call - I hadn't come across bimap but that looks like a much better solution. I've added this to #26539 to do as a separate PR because there are several similar functions already and I'd like to do them all in one go.

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/Reduction/ValidatePerThetaDefaults.cpp:
```diff
@@ -106,6 +106,16 @@ PerThetaDefaultsValidator::parseProcessingInstructions(
+PerThetaDefaultsValidator::parseBackgroundProcessingInstructions(
+    CellText const &cellText) {
+  auto optionalInstructionsOrNoneIfError = ::MantidQt::CustomInterfaces::
+      ISISReflectometry::parseProcessingInstructions(cellText[9]);
```
I agree this would be better. I have an issue to do this already as part of #26539 so will do it as a separate PR.

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/Reduction/PerThetaDefaults.cpp:
```diff
@@ -83,6 +93,8 @@ perThetaDefaultsToArray(PerThetaDefaults const &perThetaDefaults) {
   if (perThetaDefaults.processingInstructions())
     result[8] = *perThetaDefaults.processingInstructions();
+  if (perThetaDefaults.backgroundProcessingInstructions())
+    result[9] = *perThetaDefaults.backgroundProcessingInstructions();
```
As above

gemmaguest commented:
> I've also noticed a failure mode where deleting all of the rows for the per-angle defaults makes it impossible to process rows, even if the per-angle defaults is re-populated with rows. However, I don't think this is a regression and may need to be a separate ticket.

I can't reproduce this - can you describe the steps you took?

gemmaguest commented:
Just rebased on master to resolve the conflicts

gemmaguest commented:
I've just noticed that the new options are not greyed out during processing - I'll fix that now

gemmaguest commented:
The new options should now be greyed out during processing.

rprospero commented:
These are the steps to reproduce the bug I encountered.  This may be a separate ticket, but it may also just be specific to my setup:

1) Start the ISIS Reflectometry Interface
2) Set the Instrument to OFFSPEC
3) Set the Run number to 53373 and the angle to 0.35 in the first row.
3.5) Press the process button
4) Check the ADS to confirm that IvsQ₅3373 was created
5) Delete all the workspaces in the ADS
6) In the ISIS Reflectometry Interface, click the Experiment Settings tab
7) Click the first row Per Angle Defaults table
8) Press the delete key to delete the row
9) Return to the runs tab
10) Press the process button
11) Confirm that the algorithm throws an exception
12) In the ISIS Reflectometry Interface, click the Experiment Settings tab
13) Click the add row button
9) Return to the runs tab
10) Press the process button
11) Confirm that the algorithm still throws an exception

gemmaguest commented:
> These are the steps to reproduce the bug I encountered. This may be a separate ticket, but it may also just be specific to my setup:

This relates to a known issue for Offspec. We get the error `Could not find component in instrument with name 'slit1'` because we need the slits to calculate the Q resolution but they are not defined in the IDF. We get around this by setting `dQ/Q` in the defaults table so that we don't need to calculate it. By deleting the row it is no longer there so it starts looking for the slits.

We could do with a better error message here. I think a lot of our error messages are confusing because they come directly from low-level algorithms with no context about how we're using them. I'll create an issue so that we can start compiling a list of things to improve.

gemmaguest commented:
@rprospero I think I've addressed all your points (admittedly by pushing to other issues!). Could you check if you're happy with my responses and if so approve?
